### ISSUE TITLE
Large partitions read performance degraded by 6 times on `2023.1.x` compared to `2022.2.x`

### ISSUE THREAD

**Issue by vponomaryov** [scylladb/scylladb#13740]
## Issue description

- [x] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

*Describe your issue in detail and steps it took to produce it.*

## Impact

The large partitions read performance degraded by `6` times on the Intel platform and for 20% on the Arm.

## How frequently does it reproduce?

100%

## Installation details

Kernel Version: 5.19.0-1022-aws
Scylla version (or git commit hash): `2023.1.0~rc4-20230402.f90af81a13cc` with build-id `4db7799d283c7ca63c7639b04db34c301dbcfb66`

Cluster size: 4 nodes (i3en.3xlarge)

Scylla Nodes used in this run:
- longevity-large-partitions-4d-dev-db-node-174f2a23-5 (3.250.108.112 | 10.4.0.192) (shards: 10)
- longevity-large-partitions-4d-dev-db-node-174f2a23-4 (3.250.214.194 | 10.4.2.134) (shards: 10)
- longevity-large-partitions-4d-dev-db-node-174f2a23-3 (54.154.160.14 | 10.4.0.64) (shards: 10)
- longevity-large-partitions-4d-dev-db-node-174f2a23-2 (3.250.49.91 | 10.4.1.16) (shards: 10)
- longevity-large-partitions-4d-dev-db-node-174f2a23-1 (34.244.124.238 | 10.4.1.84) (shards: 10)

OS / Image: `ami-000f0d43cb22bd09e` (aws: eu-west-1)

Test: `vp-longevity-large-partition-4days-test`
Test id: `174f2a23-40fa-4e14-9f32-6649455c72ee`
Test name: `scylla-staging/valerii/vp-longevity-large-partition-4days-test`
Test config file(s):

- [longevity-large-partition-4days.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/9afbf4febc7b40de240784732a4496de7216fb8c/test-cases/longevity/longevity-large-partition-4days.yaml)

## Details:
The behavior of the `2023.1.x` is similar to the `5.2.x` ones.

`Large partitions read load` on Scylla `2023.1.0-rc4` ([CI job link](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/8/console)):
![Screenshot from 2023-04-28 18-46-31](https://user-images.githubusercontent.com/3796058/235432017-0e0a85a1-080b-4e02-be34-bc6440f77f2b.png)

`Large partitions read load` on Scylla `2022.2.5` ([CI job link](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/9/console)):
![Screenshot from 2023-05-01 11-15-16](https://user-images.githubusercontent.com/3796058/235431906-15df08b9-64eb-4e93-a1f0-2276d11bf93e.png)

So, as we can see on the pictures, the `2022.2.5` `large partitions read load` is about `6k op/s` and using `2023.1.x` it happens to reach `1k op/s` but the stable variant is even less till the `750 op/s`.

Then, `2023.1.x` on `Arm`:
![Screenshot from 2023-04-28 18-48-49](https://user-images.githubusercontent.com/3796058/235432900-dd1e37a0-4aa8-4a90-a981-a89a39f71fe5.png)

`2022.2.5` on `Arm`:
![Screenshot from 2023-05-01 11-16-00](https://user-images.githubusercontent.com/3796058/235432949-2d5636c6-add6-4f75-8651-b09eabeb88d3.png)

So, here we see the `large partitions read load` degradation for about 20%.

- Restore Monitor Stack command: `$ hydra investigate show-monitor 174f2a23-40fa-4e14-9f32-6649455c72ee`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=174f2a23-40fa-4e14-9f32-6649455c72ee)
- Show all stored logs command: `$ hydra investigate show-logs 174f2a23-40fa-4e14-9f32-6649455c72ee`

## Logs:
- **db-cluster-174f2a23.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/db-cluster-174f2a23.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/db-cluster-174f2a23.tar.gz)
- **sct-runner-events-174f2a23.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/sct-runner-events-174f2a23.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/sct-runner-events-174f2a23.tar.gz)
- **sct-174f2a23.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/sct-174f2a23.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/sct-174f2a23.log.tar.gz)
- **monitor-set-174f2a23.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/monitor-set-174f2a23.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/monitor-set-174f2a23.tar.gz)
- **loader-set-174f2a23.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/loader-set-174f2a23.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/174f2a23-40fa-4e14-9f32-6649455c72ee/20230428_114707/loader-set-174f2a23.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/8/)

**Comment by mykaul** [comment]
@vponomaryov - good finding! Anything alarming in the logs (such as stalls) that can explain this?

**Comment by mykaul** [comment]
I'm unsure if it suffers from https://github.com/scylladb/scylladb/issues/13560 - this is a smaller node, so I'm unsure, but we better re-test with a 5.15 kernel (RC5 build)

**Comment by vponomaryov** [comment]
> @vponomaryov - good finding! Anything alarming in the logs (such as stalls) that can explain this?

Haven't looked into DB logs details yet. TBD.

> I'm unsure if it suffers from #13560 - this is a smaller node, so I'm unsure, but we better re-test with a 5.15 kernel (RC5 build)

@mykaul Triggered CI jobs with the `5.2.0-rc5`.

**Comment by vponomaryov** [comment]
Scylla `5.2.0-rc5` results:

# Intel
![Screenshot from 2023-05-01 16-25-30](https://user-images.githubusercontent.com/3796058/235458035-9fe597a3-affc-4b6d-b289-302aff99bb96.png)

# Arm
![Screenshot from 2023-05-01 16-25-37](https://user-images.githubusercontent.com/3796058/235458051-572fb067-6a5e-4907-a64f-3c634d0b9aa8.png)

So, the bug is not reproduced using the `rc5` version.
Interesting to note that the `Arm` platform shows better results than `Intel` does. The diff is about `25%` .

**Comment by mykaul** [comment]
@vponomaryov - thanks for the quick turnaround! Q: if the bug doesn't reproduce, why not close it?

**Comment by michoecho** [comment]
Reopening. This is unrelated to https://github.com/scylladb/scylladb/issues/13560. You shouldn't have considered that in the first place because i3en isn't supposed to be affected, since it doesn't use Xen.

A look at the metrics shows that queries in [the 2022.2.5 run](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/9/console) are perfectly cached:
![image](https://user-images.githubusercontent.com/45052050/235694676-60f40110-7429-4497-8c53-5d15bf51a707.png)
while in the [the 2023.1.0~rc4 run](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/8/console) they are not:
![image](https://user-images.githubusercontent.com/45052050/235695197-497df152-4957-4b1b-863d-55a2746ce656.png).

Here's the weird part: the test runs
```
scylla-bench -workload=uniform -mode=read -replication-factor=3 -partition-count=60 -clustering-row-count=10000000 -clustering-row-size=2048
```
, which AFAIK means that each replica should have 600M rows, each with an equal chance of being read.

So to cache all rows perfectly, as in the "good run", the cache would have to contain 600M rows. But that's not true:
![image](https://user-images.githubusercontent.com/45052050/235698993-6c6cd96a-3619-49df-91f2-60f01a08cb4b.png)

Not only it's much less than 600M, one of the nodes only has 90k rows in the cache, yet it is still serving perfectly cached reads.

For completeness, the "bad run" looks reasonable:
![image](https://user-images.githubusercontent.com/45052050/235699681-c781da09-91df-4c72-b0ec-99f114d3b4b8.png)

From this it also appears that the maximum possible cache capacity is ~30M rows, which checks out sense since that's about equal to 70 GiB (total cache memory) / 2kiB (row size).

Looks like a scylla-bench bug to me. As if it wasn't actually generating a uniform distribution sometimes.

**Comment by michoecho** [comment]
The fact that it failed on rc4 but succeeded on rc5 is most likely a coincidence, a result of a flaky bug.

**Comment by vponomaryov** [comment]
> Looks like a scylla-bench bug to me. As if it wasn't actually generating a uniform distribution sometimes.

There were pairs (`Intel` and `Arm`) of CI job runs for each of the `3` Scylla versions - `2022.2.5`, `2023.1.0-rc4` and `2023.1.0-rc5`. 
So, `6` total jobs used in this report.
All the runs did use completely the same `scylla-bench` docker image and the setup configuration.
Both jobs of each pair were behaving alike to each other. In each pair `Arm` results were better for `25-40%` than `Intel` had. `3` pairs had this diff.

Also, we did have lots of other runs which had much more severe load.
And exactly `2023.1.x-x` were failing even to survive the load values which we had for long time.
So, I started reducing the load values to find a suitable enough values to have non-timing out query results. 
Then, I ran many more CI jobs when was polishing the load command values.

Also, I tried pretty old `scylla-bench` for comparison which was known to work `ok enough` with the `2022` and it also failed on the `2023`.
I have no evidence which commit(s) exactly fixed the problem but the problem definitely gone with `rc5`.

So, all of the above facts make me strongly disagree that it is `flaky` and `scylla-bench`'s bug.

Just for case the `2023.1.0-rc5` job results are needed:
`Intel`: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/10/console
`Arm`: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-arm-test/6/console

**Comment by michoecho** [comment]
> I have no evidence which commit(s) exactly fixed the problem but the problem definitely gone with rc5.

@vponomaryov Scylla's RAM can only fit ~30M rows. If scylla-bench is working as expected, (uniform distribution on 600M rows), it should be *mathematically impossible* to reach cache hit ratio greater than ~5%. But 2022.2.5 and rc5 reach 100%.

Do you see what I'm saying? I'm claiming that all runs are behaving weird, not just rc4. And in fact rc4 seems the closest to what is mathematically expected from the test.

> Both jobs of each pair were behaving alike to each other.

I agree it's surprising and unlikely for the kind of flaky bug I'm suggesting to manifest this way.
But I'm not claiming there is no connection between the Scylla version and the behavior of the test.
I'm claiming that the results for "good" runs seem less expected than the ones for the "bad" runs,
so it doesn't look like a temporary regression in rc4 to me.

> There are 3 of them, running the same load from the same point.
So, 2 from 3 of the same queries are capable of getting just cached value from the first-processed query from the third loader.

Are you sure? A quick look at scylla-bench's code says to me that the randomness there is seeded with `time.Now().UnixNano() + int64(os.Getpid())`. So loaders shouldn't be following the same sequence of reads unless all loader machines start with exactly the same timestamp, which seems rather unlikely.

But even if that happens, that should only scale the miss ratio by 1/3. So if all loaders followed the same sequence, it should be impossible to reach >70% hit ratio, rather than a >5% hit ratio. It should still impossible to reach 100%.

**Comment by vponomaryov** [comment]
> > I have no evidence which commit(s) exactly fixed the problem but the problem definitely gone with rc5.
> 
> @vponomaryov Scylla's RAM can only fit ~30M rows. If scylla-bench is working as expected, (uniform distribution on 600M rows), it should be _mathematically impossible_ to reach cache hit ratio greater than ~5%. But 2022.2.5 and rc5 reach 100%.

Cache is mutable. 5% is part we can keep in cache at the same time, but it is mutable.

> > Both jobs of each pair were behaving alike to each other.
> 
> I agree it's surprising and unlikely for the kind of flaky bug I'm suggesting to manifest this way. But I'm not claiming there is no connection between the Scylla version and the behavior of the test. I'm claiming that the results for "good" runs seem less expected than the ones for the "bad" runs, so it doesn't look like a temporary regression in rc4 to me.

It is not only `rc4`. It was hit also in previous rc's.

> > There are 3 of them, running the same load from the same point.
> > So, 2 from 3 of the same queries are capable of getting just cached value from the first-processed query from the third loader.
> 
> Are you sure? A quick look at scylla-bench's code says to me that the randomness there is seeded with `time.Now().UnixNano() + int64(os.Getpid())`. So loaders shouldn't be following the same sequence of reads unless all loader machines start with exactly the same timestamp, which seems rather unlikely.
> 
> But even if that happens, that should only scale the miss ratio by 1/3. So if all loaders followed the same sequence, it should be impossible to reach >70% hit ratio, rather than a >5% hit ratio. But it should still be impossible to reach 100%.

The random values are applied only for the `workload` part, but the `partition keys` and `clustering keys` get incremented by 1 all the time. It may be in `asc` or `desc` order, but we use the default `asc` one on all the loaders.
The randomness is applied only in scope of the `10Gb` data size of each of the linary read partitions.
So, we walk through the same partitions on each of the 3 loaders all the time.

This situation doesn't make me think that we should stick to the total amount of data which is `600Gb` calculating `mathematical` limit.

**Comment by michoecho** [comment]
> Cache is mutable. 5% is part we can keep in cache at the same time, but it is mutable.

Yes, but why are you pointing this out? In a uniform distribution the state of the cache doesn't matter, only its size.

> It is not only rc4. It was hit also in previous rc's.

Do we know when it was hit for the first time?

> The random values are applied only for the workload part, but the partition keys and clustering keys get incremented by 1 all the time.

I don't understand what incrementation you are talking about. As far as I can see, the scylla-bench invocation used in this test repeatedly sends queries of the form:
```
SELECT pk, ck, v FROM scylla_bench.test WHERE pk = ? AND ck >= ?  LIMIT 2000
```
(https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/modes.go#L501),
where `pk` and `ck` are independently sampled from the uniform distribution for each query
(
https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/modes.go#L510
https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/modes.go#L529
https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/pkg/workloads/workloads.go#L106
https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/pkg/workloads/workloads.go#L99
).

> The randomness is applied only in scope of the 10M rows of each of the linary read partitions.
It is 3 times less than the capacity of the cache.
So, we walk through the same partitions on each of the 3 loaders all the time.

> This situation doesn't make me think that we should stick to the total amount of data which is 600M rows calculating mathematical limit.

I don't understand. Could you rephrase that?

**Comment by vponomaryov** [comment]
> > Cache is mutable. 5% is part we can keep in cache at the same time, but it is mutable.
> 
> Yes, but why are you pointing this out? In a uniform distribution the state of the cache doesn't matter, only its size.

I tried to say that pointing out at the `5%` saying it is not possible, with this setup, to have more than that  is not correct from my POV.

> > It is not only rc4. It was hit also in previous rc's.
> 
> Do we know when it was hit for the first time?

I know which we observed first. We do not run this test often enough.
`2023.1.0~rc2-20230302.726f8a090337`
`5.2.0~rc2-20230228.908a82bea064`

> > The random values are applied only for the workload part, but the partition keys and clustering keys get incremented by 1 all the time.
> 
> I don't understand what incrementation you are talking about. As far as I can see, the scylla-bench invocation used in this test repeatedly sends queries of the form:
> 
> ```
> SELECT pk, ck, v FROM scylla_bench.test WHERE pk = ? AND ck >= ?  LIMIT 2000
> ```
> 
> (https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/modes.go#L501), where `pk` and `ck` are independently sampled from the uniform distribution for each query ( https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/modes.go#L510 https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/modes.go#L529 https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/pkg/workloads/workloads.go#L106 https://github.com/scylladb/scylla-bench/blob/0c8d921206726e0738b115f1a3190959789c6f92/pkg/workloads/workloads.go#L99 ).
> 
> > The randomness is applied only in scope of the 10M rows of each of the linary read partitions.
> > It is 3 times less than the capacity of the cache.
> > So, we walk through the same partitions on each of the 3 loaders all the time.
> 
> > This situation doesn't make me think that we should stick to the total amount of data which is 600M rows calculating mathematical limit.
> 
> I don't understand. Could you rephrase that?

About the `s-b` random approach I was mistaken and looked at wrong place in the code. 
You are right about the approach for getting the pks and cks in this particular setup.
But it doesn't help anyhow in current bug context.

`2023.1.0-rc5` doesn't have the perf degradation compared to the `2022.2.5`.
I ran one more CI job and it passed as expected: https://jenkins.scylladb.com/job/scylla-staging/job/valerii/job/vp-longevity-large-partition-4days-test/11/console
All 3 runs of the `2023.1.0-rc5` I have made so far show a bit better perf than `2022.2.5` has using the same setup.
And testing history shows that rc2-rc4 definitely were affected by this bug.
I am not aware about cases where `rc1` or earlier versions were tested using this setup.

So, as a result, it is unclear to me why this bugreport was reopened when it's main idea of the `perf degradation` is not in place as of the `rc5`.

**Comment by michoecho** [comment]
> So, as a result, it is unclear to me why this bugreport was reopened when it's main idea of the perf degradation is not in place as of the rc5.

1. Because we are supposed to understand issues before we close them. When a symptom disappears without explanation, that doesn't mean the issue was solved.
2. And, more importantly, because the "main idea" appears invalid. It's not good that the test went back to the usual behavior if the usual behavior was wrong in the first place.

I repeat: the test is meant to generate a random read workload with a uniform distribution over 600 M rows, with a cache size of 30 M rows. It should be impossible in this workload to reach a cache hit ratio of 100%. And yet that's what we see, and the issue considers that the expected behavior. How come?

**Comment by roydahan** [comment]
Here is the list of commits between RC4 and RC5, maybe something would pop up:

```
* | | a47bcb26e4 - (tag: scylla-2023.1.0-rc5) release: prepare for 2023.1.0-rc5 (4 days ago) <Yaron Kaikov>
* | |   b5e4f65913 - Merge branch 'branch-5.2' of github.com:scylladb/scylladb into next-2023.1 (6 days ago) <Avi Kivity>
|\ \ \
| | |/
| |/|
| * | 429b696bbc - release: prepare for 5.2.0 (6 days ago) <Beni Peled>
| * | a89867d8c2 - release: prepare for 5.2.0-rc5 (8 days ago) <Beni Peled>
| * | 6ad94fedf3 - utils: clear_gently: do not clear null unique_ptr (9 days ago) <Benny Halevy>
| * | 50095cc3a5 - Merge 'db: system_keyspace: use microsecond resolution for group0_history range tombstone' from Kamil Braun (10 days ago) <Botond Dénes>
| * |   7b2215d8e0 - Merge 'Backport bugfixes regarding UDT, UDF, UDA interactions to branch-5.2' from Wojciech Mitros (2 weeks ago) <Botond Dénes>
| |\ \
| | * | 7242c42089 - uda: change the UDF used in a UDA if it's replaced (2 weeks ago) <Wojciech Mitros>
| | * | 70ff69afab - functions: add helper same_signature method (2 weeks ago) <Wojciech Mitros>
| | * | 5fd4bb853b - uda: return aggregate functions as shared pointers (2 weeks ago) <Wojciech Mitros>
| | * | 313649e86d - udf: also check reducefunc to confirm that a UDF is not used in a UDA (2 weeks ago) <Wojciech Mitros>
| | * | 14d8cec130 - udf: fix dropping UDFs that share names with other UDFs used in UDAs (2 weeks ago) <Wojciech Mitros>
| | * | 203cbb79a1 - pytest: add optional argument for new_function argument types (2 weeks ago) <Wojciech Mitros>
| | * | 51f19d1b8c - udt: disallow dropping a user type used in a user function (2 weeks ago) <Wojciech Mitros>
| * | | da9f90362d - Merge 'Compaction reevaluation bug fixes' from Raphael "Raph" Carvalho (2 weeks ago) <Botond Dénes>
| * | | c9a17c80f6 - mutation/mutation_compactor: consume_partition_end(): reset _stop (2 weeks ago) <Botond Dénes>
| * | |   9d384e3af2 - Merge 'Backport "reader_concurrency_semaphore: don't evict inactive readers needlessly" to branch-5.2' from Botond Dénes (2 weeks ago) <Avi Kivity>
| |\ \ \
| | * | | 3e10c3fc89 - reader_concurrency_semaphore: don't evict inactive readers needlessly (3 weeks ago) <Botond Dénes>
| | * | | f11deb5074 - reader_concurrency_semaphore: add stats to record reason for queueing permits (3 weeks ago) <Botond Dénes>
| | * | | 1baf9dddd7 - reader_concurrency_semaphore: can_admit_read(): also return reason for rejection (3 weeks ago) <Botond Dénes>
| * | | | 0da0c94f49 - cql: USING TTL 0 means unlimited, not default TTL (2 weeks ago) <Nadav Har'El>
| * | | | 1a9f51b767 - cql: fix empty aggregation, and add more tests (2 weeks ago) <Nadav Har'El>
| * | | | dba0e604a7 - table: Fix disk-space related metrics (2 weeks ago) <Raphael S. Carvalho>
| | |/ /
| |/| |
* | | |   09557493d9 - Merge 'Next 2023.1 backport per opcode metrics' from Vladislav Zolotarov (12 days ago) <Botond Dénes>
|\ \ \ \
| |_|_|/
|/| | |
| * | | 8f1cf11aee - transport: add per-scheduling-group CQL opcode-specific metrics (13 days ago) <Vlad Zolotarov>
| * | | cd1a1b90ef - transport: refactor CQL metrics (13 days ago) <Vlad Zolotarov>
|/ / /
* | | 4e025929af - Merge branch 'branch-5.2' of github.com:scylladb/scylladb into next-2023.1 (2 weeks ago) <Avi Kivity>
|\| |
| * | 4ea67940cb - locator: token_metadata: get rid of a quadratic behaviour in get_address_ranges() (2 weeks ago) <Michał Chojnowski>
| * | a8c49c44e5 - cql/query_options: add a check for missing bind marker name (2 weeks ago) <Jan Ciolek>
| * | 12a29edf90 - test/alternator: fix flaky test for partition-tombstone scan (3 weeks ago) <Nadav Har'El>
| |/
| * 128050e984 - Merge 'commitlog: Fix updating of total_size_on_disk on segment alloc when o_dsync is off' from Calle Wilund (4 weeks ago) <Botond Dénes>
| * d70751fee3 - release: prepare for 5.2.0-rc4 (4 weeks ago) <Yaron Kaikov>
* |   ddb6cced34 - Merge 'Replicated provider - initialize tables on startup (backport for 2023.1)' from Calle Wilund (2 weeks ago) <Botond Dénes>
|\ \
| * | ae0b311d13 - replicated_provider: Initialize tables on system start (2 weeks ago) <Calle Wilund>
| * | b569fe3614 - alternator/rest api tests: Remove name assumption and rely on actual scylla info (2 weeks ago) <Calle Wilund>
| * | c0333a0780 - init: Add life cycle notifications to configurables (2 weeks ago) <Calle Wilund>
* | | 4ff2820a7c - audit: do not round time (3 weeks ago) <Juliusz Stasiewicz>
|/ /
* | f90af81a13 - (tag: scylla-2023.1.0-rc4) release: prepare for 2023.1.0-rc4 (4 weeks ago) <Yaron Kaikov>
```

**Comment by roydahan** [comment]
> > So, as a result, it is unclear to me why this bugreport was reopened when it's main idea of the perf degradation is not in place as of the rc5.
> 
> 1. Because we are supposed to understand issues before we close them. When a symptom disappears without explanation, that doesn't mean the issue was solved.
> 2. And, more importantly, because the "main idea" appears invalid. It's not good that the test went back to the usual behavior if the usual behavior was wrong in the first place.
> 
> I repeat: the test is meant to generate a random read workload with a uniform distribution over 600 M rows, with a cache size of 30 M rows. It should be impossible in this workload to reach a cache hit ratio of 100%. And yet that's what we see, and the issue considers that the expected behavior. How come?

The test doesn't implicitly check for the cache hit ratio, but you're right that if there is something odd here, we need to look into it.
@vponomaryov I know that you're running things from your own branch here, so let's make sure that what you run and what's @michoecho is looking at are the same things in terms of the test configuration.

**Comment by michoecho** [comment]
> The test doesn't implicitly check for the cache hit ratio

Doesn't check for it explicitly, you mean? Every performance test implicitly checks for the cache hit ratio. The "performance degradation" reported in the issue is all about the cache hit ratio being lower than it used to.

> @vponomaryov I know that you're running things from your own branch here, so let's make sure that what you run and what's @michoecho is looking at are the same things in terms of the test configuration.

I'm not looking at the source code but at the invocations I see in Jenkins logs for the relevant runs. So there should be no problem with me looking at the wrong code.

**Comment by avikivity** [comment]
Does the test perform scans of large partitions? Paged reads could be impacted by changes to reader_concurrency_semaphore.

Were the tests on 2022.2 conducted with the index cache enabled or disabled? We only disabled it in rc6. If the test was run before then, then it would be with the index cache enabled (and would have had better performance for large partitions).

**Comment by vponomaryov** [comment]
> Does the test perform scans of large partitions? Paged reads could be impacted by changes to reader_concurrency_semaphore.
> 

In all the test runs used in this bugreport the scans were disabled.

> Were the tests on 2022.2 conducted with the index cache enabled or disabled? We only disabled it in rc6. If the test was run before then, then it would be with the index cache enabled (and would have had better performance for large partitions).

All the test runs used here were run recently using the same setup configuration.
From `2022` branch was used `2022.2.5`.
We do not change the cache configuration in this setup.
So, it must be the default value in each of the test runs used here.

**Comment by michoecho** [comment]
I re-run the workload and debugged it, and everything is as I said. (As it should, because it was mathematically the only possibility).

It has nothing to with Scylla. The problem is that scylla-bench doesn't actually generate the expected uniform workload in some runs.
Except the bug is not in scylla-bench itself, but in how SCT uses it.

scylla-bench sends random queries of the form `SELECT pk, ck, v FROM scylla_bench.test WHERE pk = ? AND ck >= ?  LIMIT 2000`. The intention is for this to be a random uniform workload.

But SCT sometimes deliberately messes up `scylla-bench`'s assumptions by manually reversing the clustering order of `scylla_bench.test` with a probability of 0.5. (https://github.com/scylladb/scylla-cluster-tests/blob/1d31a3941d1ba6c9ab69bc16f603aa1d36b70505/test_lib/scylla_bench_tools.py#L13. What's the point of doing that, anyway?)

The workload is uniform when the clustering order is the default ASC.
But when the order is changed to DESC, the above `SELECT` stops being uniform as intended — it will just always read the same first 2000 rows of the partition. (The 9999999-9998000 clustering range), which is 100% cacheable, and the workload is easier by an order of magnitude.

>  I have no evidence which commit(s) exactly fixed the problem but the problem definitely gone with rc5.

@vponomaryov It turns out to be a pure coincidence after all. Repeating a test 6 times is not enough to say that something is "definitely" true.

**Comment by vponomaryov** [comment]
> @vponomaryov It turns out to be a pure coincidence after all. Repeating a test 6 times is not enough to say that something is "definitely" true.

Actually 7. The probability of this case is less than 1%. Probably, I need to buy a lottery ticket.
And only [8th run](https://argus.scylladb.com/test/eaaeef08-731f-4c19-bcbf-4547ccdacf1b/runs?additionalRuns[]=1a262665-2082-493b-b48e-6f055c49f7b9) using `rc5` went out of the `coincidence chain` showing different results than other runs.

So, it is SCT bug. This one can be closed
@michoecho , Thanks!

**Comment by roydahan** [comment]
@yarongilor The bug here is that after you removed the reversed query load from this scenario, you forgot to remove also the randomness of the schema creation.

**Comment by avikivity** [comment]
> I re-ran the workload and debugged it, and everything is as I said. (As it should, because it was mathematically the only possibility).
> 
> It has nothing to with Scylla. The problem is that scylla-bench doesn't actually generate the expected uniform workload in some runs. Except the bug is not in scylla-bench itself, but in how SCT uses it.
> 
> scylla-bench sends random queries of the form `SELECT pk, ck, v FROM scylla_bench.test WHERE pk = ? AND ck >= ? LIMIT 2000`. The intention is for this to be a random uniform workload.
> 
> But SCT sometimes deliberately messes up `scylla-bench`'s assumptions by manually reversing the clustering order of `scylla_bench.test` with a probability of 0.5. (https://github.com/scylladb/scylla-cluster-tests/blob/1d31a3941d1ba6c9ab69bc16f603aa1d36b70505/test_lib/scylla_bench_tools.py#L13. What's the point of doing that, anyway?)
> 
> The workload is uniform when the clustering order is the default ASC. But when the order is changed to DESC, the above `SELECT` stops being uniform as intended — it will just always read the same first 2000 rows of the partition, (the [9999999](https://github.com/scylladb/scylladb/commit/9999999dab2dc8c32026c25f826a1a146e33b8af)-9998000 clustering range), which is 100% cacheable, and the workload is easier by an order of magnitude.
> 
> > I have no evidence which commit(s) exactly fixed the problem but the problem definitely gone with rc5.
> 
> @vponomaryov It turns out to be a pure coincidence after all. Repeating a test 6 times is not enough to say that something is "definitely" true.

Good detective work.

**Comment by yarongilor** [comment]
> @yarongilor The bug here is that after you removed the reversed query load from this scenario, you forgot to remove also the randomness of the schema creation.

@roydahan , i think the bug here was much simpler than requiring @vponomaryov 's fix PR.
We first edited 4d yaml, then rollback all changes since introduced 2d yaml.
But if this line wasn't rolled-back, the issue wouldn't have happened:
```
# Set cql_schema_seed in order to get: "CLUSTERING ORDER BY (ck ASC)"
cql_schema_seed: '930'
```
